### DISCUSSION TITLE
How to learn the usage of specified feature in Argo Workflow?

### DISCUSSION THREAD

**Discussion by link89** [argoproj/argo-workflows#12608] [Q&A] [answered]
I am asked to work on a project that use Argo Workflow for scientific computing recently.  It's easy to run some sample workflow at the beginning, but I find that I cannot figure out how specific feature works when I start a real-world project.

For example, a recent question I run into is "Can I use `arguments.artifacts`  to set a s3 key for a key only output artifacts? "

All the documents and examples I can find are
* https://argo-workflows.readthedocs.io/en/latest/fields/#workflowstep
* https://argo-workflows.readthedocs.io/en/latest/walk-through/steps/
* https://github.com/argoproj/argo-workflows/blob/main/examples/steps.yaml

And I still confuse after reading them. Is there better way for me to learn features and design ideas of Argo workflow. Or do I have to dive into the source code to figure them out?

**Comment by agilgur5** [comment, marked as answer]
> Or do I have to dive into the source code to figure them out?

No, you definitely shouldn't need to do that. I was able to do most things without looking at the source, and only needed to check the source for debugging, bugfixes, and new features.
Argo has a ton of options though, so sometimes a very specific configuration is not covered well in the docs or hard to discover. PRs very welcome to improve the docs -- many of my own PRs are to the docs.

> All the documents and examples I can find are

It seems like you're looking at `steps`, but your question is more specifically about `artifacts` -- is there a reason you didn't look at the Artifacts docs?

For instance:
- https://argo-workflows.readthedocs.io/en/latest/walk-through/artifacts/
- https://argo-workflows.readthedocs.io/en/latest/workflow-inputs/
- https://argo-workflows.readthedocs.io/en/latest/key-only-artifacts/
- https://argo-workflows.readthedocs.io/en/latest/fields/#artifact

> For example, a recent question I run into is "Can I use `arguments.artifacts` to set a s3 key for a key only output artifacts? "

Also, in general, with any questions, it's good to include a Workflow that you tried or what you expected to work that didn't.
Writing and testing workflows is also a great way to learn and become more confident in your knowledge.
